docs/dev/guidelines/server.rst

```diff
@@ -350,3 +350,240 @@ Please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-c
+        TranslationServer testTranslationServer = new TranslationServer() {
+            public String translate(String word, String language) {
+                return word + ":" + language;
+            }
+        }
+        String expected = "Dog:French";
+        String actual = testTranslationServer.runService("Dog", "French");
+        if(!expected.equals(actual)) fail("Descriptive fail message");
+    }
+
+This test correctly tests if the student added the case for the French language and called the appropriate method with the appropriate parameters. Because the translation method was overridden,
+it doesn't matter whether the student has already completed the previous task or not.
+(If you do that, you should have some way to deal with students that make the class or method final, either via problem statement or test, otherwise students get compilation errors in the test code)
+
+
+Make use of JUnit 5 Features:
+https://junit.org/junit5/docs/current/user-guide/#writing-tests
+https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html
+
+General Testing best practices are found on https://phauer.com/2019/modern-best-practices-testing-java/
+
+Also check out https://www.baeldung.com/spring-tests
+
+And https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/
+
+
+23. Avoid using @MockBean
+=========================
+
+Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here. <https://www.baeldung.com/spring-tests>`_
+Basically every time @MockBean appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
+This leads to a large overhead, which makes the tests tend to take a lot more time.
+
+Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.
+
+.. code-block:: java
+
+    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+        @SpyBean
+        private FileUploadSubmissionExportService fileUploadSubmissionExportService;
+
+        @Test
+        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+        public void testExportAll_IOException() throws Exception {
+            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
+            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
+        }
+    }
+
+This, as mentioned above, we should really avoid this.
+Instead we can use Static Mocks. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.
```

Comment:
Maybe a link to more information about static mocks would be helpful?

Reply:
```suggestion
Instead we can use `Static Mocks <https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/>`_. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.
```